### PR TITLE
[charts] Support BarChart with `Date` data

### LINKED ISSUE #12537: [Charts] Support BarChart with scaleType: time
### Summary

Trying out charts on a small weather dashboard. Would like to show precipitation over time in a bar chart. See bottom chart on https://dashboard-framework-prototype.vercel.app/forecast. Would like the X-axis to look like the other two charts.

Would it make sense to add a `bars: true` to the `LineChart` series, analog to the `area: true`?

### Examples

_No response_

### Motivation

_No response_

**Search keywords**: barchart time

### LINKED ISSUE #13046: [charts] Support for ordinal axes
## Summary

Ordinal axes are essential for visualizing data where one of the axis have a specific order but don't represent continuous values (e.g., weekdays, months, quarters).

Ordinal axes prioritize order over uniform spacing. This means data points appear at consistent intervals on the chart regardless of the actual time or distance between them. Consequently, missing data periods, like nights or weekends, won't cause gaps in the visual representation.

E.g. Stock charts usually are usually ordinal by default.

## Motivation

- Enables the visualization of a wider range of data types, including time series data, survey responses, and other categorical information.
- Improves the accuracy and clarity of charts by ensuring proper representation of ordered categories.
- Enhances user experience by providing a more intuitive way to understand the relationships between data points.

## Benchmark

- Highcharts [Example](https://www.highcharts.com/samples/stock/xaxis/ordinal-true) | [API](https://api.highcharts.com/highstock/xAxis.ordinal) 
- https://charts.ag-grid.com/react/axes-types/#ordinal-time

**Search keywords**:

### PR DESCRIPTION
**Pull request by alexfauquette**
@JCQuintas I was looking at your PR #13454 and I thought about another way of doing it

https://codesandbox.io/p/sandbox/zealous-morning-9lw243?file=%2Fsrc%2FDemo.tsx%3A5%2C1

This PR try to go the other way.

- In #13454 the Bar plot is adapted to support the time scale
- In this one we create a new axis type that generate a band scale manipulating date. We could also make it kind of magic by keeping band type and just checking if the first data element is a `Date`

fixes #12537
fixes #13046

### Changelog

The band scale with `Date` data get a better default formater

### CODE REVIEW

File: packages/x-charts/src/context/CartesianContextProvider.tsx [hunk 1]
```diff
@@ -224,6 +224,16 @@ function CartesianContextProvider(props: CartesianContextProviderProps) {
...
         };
+        if (axis.scaleType === 'ordinal-time') {
+          const timeScale = scaleTime(axis.data!, range);
+          completedXAxis[axis.id].valueFormatter =
+            axis.valueFormatter ??
+            ((v, { location }) =>
+              location === 'tick'
+                ? timeScale.tickFormat(axis.tickNumber)(v)
+                : `${v.toLocaleString()}`);
+          // completedXAxis[axis.id].tickInterval = timeScale.ticks(axis.tickNumber);
```
Comment (alexfauquette) [inline packages/x-charts/src/context/CartesianContextProvider.tsx:235]: Need to be sure the generated tick exist in the data

File: packages/x-charts/src/context/CartesianContextProvider.tsx [hunk 2]
```diff
@@ -224,6 +224,16 @@ function CartesianContextProvider(props: CartesianContextProviderProps) {
               ? getOrdinalColorScale({ values: axis.data, ...axis.colorMap })
               : getColorScale(axis.colorMap)),
         };
+        if (axis.scaleType === 'ordinal-time') {
```
Comment (alexfauquette) [inline packages/x-charts/src/context/CartesianContextProvider.tsx:227]: We could skip the craetion of the special scaletype with
```suggestion
        if (axis.data[0] instanceof Date) {
```

But that would add some kind of magic

**Comment by JCQuintas** [comment]
🤔 I think it can make a lot of sense to create this as a separate axis type, but I would probably want it also respect missing data points https://codesandbox.io/p/sandbox/zealous-morning-9lw243

It should behave like:
Screenshot 2024-06-13 at 16 27 44

Instead of:
Screenshot 2024-06-13 at 16 27 58

We could push that to the user instead too, to fill the gap, but I don't think that is a good idea.

**Comment by alexfauquette** [comment]
> I would probably want it also respect missing data points

You can have missing data point by providing `null` values. But you need to provide the `x` value.
https://mui.com/x/react-charts/lines/#skip-missing-points

image

Respecting missing data point in time axis is a bit complex, because it's not well defined.

Of course, if you have 
- `1, 2, 3, 4, 10, 11, 12` you can expect to have a bar with width 1
- `2, 4, 6, 10, 12`  you can expect to have a bar with width 2
- `2, 3, 4, 6, 10, 12`  you should probably have bar of unit 1 but with lots of holls

To do so we will have to compute what is the minimal distance between two consecutive

**Comment by JCQuintas** [comment]
@alexfauquette yeah my idea would be for the users to provide either a function or a `time|hour|minute|...` string we can generate the correct data.

**Comment by alexfauquette** [comment]
I was think about something like

```jsx
<BarChart
	xAxis={[{
		scale: 'ordinal-time',
		data: timeGenerator({
			from: new Date(2023, 5, 10),
			to: new Date(2023, 6, 15),
			step: 24 * 3600 * 1000
		})
	{/* ... */}
/>
```

But not convinced it's necessary. I assume if you have some data at regular intervals, you also have the time stamps.

@Janpot do you have some insight on this aspect: how to handle missing, or irregular time stamp for bar chart ?

**Comment by Janpot** [comment]
I assume the common use case would be grouped/aggregated data, e.g. revenue per day, or # of errors per hour. So fixed intervals, where missing data means 0, no bar.

**Comment by JCQuintas** [comment]
> I assume the common use case would be grouped/aggregated data, e.g. revenue per day, or # of errors per hour. So fixed intervals, where missing data means 0, no bar. I'm not really sure how I'd translate that into an API

If you are consuming APIs you probably have these options. But real time data, as logs and IoT wouldn't have it in a consistent way.

We could always just add this and wait for extra input there too 😅

**Comment by Janpot** [comment]
> But real time data, as logs and IoT wouldn't have it in a consistent way.

But how do you show raw realtime data in a bar chart? You'd always do some sort of bucketing in fixed intervals. What else would be the meaning of "a bar" on a time scale? Do you have a real world use-case of a bar chart with a time scale where the bars don't represent the same interval?

**Comment by JCQuintas** [comment]
@Janpot yeah, I agree with what you are saying, my point was more if we should provide the user with the means to aggregate the data and display it, or if the user should handle that themselves.

**Comment by Janpot** [comment]
> @Janpot yeah, I agree with what you are saying, my point was more if we should provide the user with the means to aggregate the data and display it, or if the user should handle that themselves.

👍  got it. Perhaps it could be valuable to have built in aggregation in the "dataSet+dataKey" part of the API. And leave the option for raw definition in the "data" part of the API. To sort of provide the option for both a low and high level abstraction. Not sure, feels like something I would use. Defining the data as an aggregation feels like a more error-tolerant way of dealing with duplicate or missing X values, it would become a non-issue.

File: packages/x-charts/src/context/CartesianContextProvider.tsx [hunk 3]
```diff
@@ -236,6 +245,15 @@ function CartesianContextProvider(props: CartesianContextProviderProps) {
               ? getOrdinalColorScale({ values: axis.data, ...axis.colorMap })
               : getColorScale(axis.colorMap)),
         };
+        if (axis.data?.[0] instanceof Date) {
```
Comment (JCQuintas) [inline packages/x-charts/src/context/CartesianContextProvider.tsx:248]: Would a `isDateData` function work here?

```tsx
const isDateData = (data?: any[]): data is Date[] => data?.[0] instanceof Date
```

File: packages/x-charts/src/context/CartesianContextProvider.tsx [hunk 4]
```diff
@@ -236,6 +245,15 @@ function CartesianContextProvider(props: CartesianContextProviderProps) {
...
+          const timeScale = scaleTime(axis.data!, range);
+          completedXAxis[axis.id].valueFormatter =
+            axis.valueFormatter ??
+            ((v, { location }) =>
+              location === 'tick'
+                ? timeScale.tickFormat(axis.tickNumber)(v)
+                : `${v.toLocaleString()}`);
```
Comment (JCQuintas) [inline packages/x-charts/src/context/CartesianContextProvider.tsx:255]: We could probably also create a `dateValueFormatter` and just `completedXAxis[axis.id].valueFormatter = dateValueFormatter(...)`

**Review by JCQuintas — APPROVED** [review]

**Comment by oscar-b** [comment]
@alexfauquette Hi! Was this change reverted? Using version `latest` in your linked sandbox doesn't work: https://codesandbox.io/p/sandbox/distracted-nightingale-rqcznv?file=/src/Demo.tsx:6,15

Or can this be achieved some other way now? Also `scaleType: "ordinal-time"` doesn't seem to exist? I'm so confused.

> MUI X: The x-axis with id "defaultized-x-axis-0" should be of type "band" to display the bar series of id "auto-generated-id-0".

**Pull request merged**